crates/ecs/src/filter.rs

```diff
@@ -0,0 +1,345 @@
+    }
+}
+
+/// A query filter that matches any entity without a given component type.
+///
+/// # Example
+/// ```
+/// # use ecs::filter::{With, Without};
+/// # use ecs::Read;
+/// # #[derive(Debug)]
+/// # struct Position;
+/// # struct Player;
+/// fn non_player_position(position: Read<Position>, _: Without<Player>) {
+///     println!("A non-player entity is at position {:?}.", position);
+/// }
+/// ```
+#[derive(Debug)]
+pub struct Without<Component: 'static> {
+    phantom: PhantomData<Component>,
+}
+
+impl<Component: Send + Sync + 'static + Sized> SystemParameter for Without<Component> {
+    type BorrowedData<'components> = ReadComponentVec<'components, Component>;
+
+    fn borrow(world: &World) -> Self::BorrowedData<'_> {
+        world.borrow_component_vec::<Component>()
+    }
+
+    unsafe fn fetch_parameter(
+        borrowed: &mut Self::BorrowedData<'_>,
+        entity: Entity,
+    ) -> Option<Self> {
+        if let Some(component_vec) = borrowed {
+            if let Some(Some(_)) = component_vec.get(entity.id) {
+                return None;
+            }
+        }
+        Some(Self {
+            phantom: Default::default(),
+        })
+    }
+
+    fn component_accesses() -> Vec<ComponentAccessDescriptor> {
+        vec![ComponentAccessDescriptor::Read(TypeId::of::<Component>())]
+    }
+}
+
+macro_rules! binary_filter_operation {
+    ($name:ident, $op:tt, $op_name:literal, $op_name_lowercase:literal) => {
+        /// A query filter that combines two filters using the `
+		#[doc = $op_name]
+		///` operation.
+        ///
+        #[doc = concat!(
+            "# Example\n```\n",
+            "# use {ecs::filter::With, ecs::filter::", stringify!($name), "};\n",
+            "# use ecs::Read;\n",
+            "# #[derive(Debug)]\n",
+            "# struct Position;\n",
+            "# struct Player;\n",
+            "# struct Enemy;\n",
+            "fn player_", $op_name_lowercase, "_enemy(position: Read<Position>, _: ", stringify!($name), "<With<Player>, With<Enemy>>) {\n",
+            "    println!(\"An entity at position {:?} is a player ", $op_name_lowercase ," an enemy.\", position);\n",
+            "}\n```",
+        )]
+        #[derive(Debug)]
+        pub struct $name<L: SystemParameter, R: SystemParameter> {
+            left: PhantomData<L>,
+            right: PhantomData<R>,
+        }
+
+        impl<L: SystemParameter, R: SystemParameter> SystemParameter for $name<L, R> {
+            type BorrowedData<'components> = (
+                <L as SystemParameter>::BorrowedData<'components>,
+                <R as SystemParameter>::BorrowedData<'components>,
+            );
+
+            fn borrow(world: &World) -> Self::BorrowedData<'_> {
+                (
+                    <L as SystemParameter>::borrow(world),
+                    <R as SystemParameter>::borrow(world),
+                )
+            }
+
+            unsafe fn fetch_parameter(
```

Comment:
The function is unsafe because it is using the same trait (`SystemParameter`) as `Read`/`Write`. And yes, there is no unsafe code in the filter implementation. To follow Rust’s guidelines, the function should be marked as unsafe because the compiler cannot guarantee the safety requirements. You could for example write this code without any compiler errors:

```rust
let mut borrowed = <P as SystemParameter>::borrow(&world);
let parameter = <P as SystemParameter>::fetch_parameter(&mut borrowed, entity);
drop(borrowed);
self(parameter);
```

Reply:
Right -- but that would only be unsafe when the `SystemParameter` is _not_ a filter-parameter, right? 

My point here is that maybe we shouldn't mark the trait as unsafe, but rather the specific implementation that actually _is_ unsafe. So maybe it shouldn't use the same trait as the unsafe version? Or the other way around, maybe there should be an `UnsafeSystemParameter`? Would that create a lot of duplicated code, do you think, or is it possible?

Reply:
The problem is that all types of system parameters need to use the `SystemParameter` trait because Rust macros cannot work a semantic level (like C++ templates). This is also the reason why `Read`/`Write` needs to be generalized to use the same trait. There are some possible workarounds to get rid of the unsafe functions in `ecs::filter`, for example:
```rust
trait SystemParameter: {
    unsafe fn unsafe_fetch_parameter(
        borrowed: &mut Self::BorrowedData<'_>,
        entity: Entity,
    ) -> Option<Self> {
        Self::fetch_parameter(borrowed, entity)
    }

    fn fetch_parameter(
        borrowed: &mut Self::BorrowedData<'_>,
        entity: Entity,
    ) -> Option<Self> {
        unreachable!()
    }
}
```
With this workaround, all safe implementations could use `fetch_parameter`, while `Read`/`Write` could override the default implementation for `unsafe_fetch_parameter`. `unsafe_fetch_parameter` is then used when running the system. But I find workaround to be clunky and it is adding unnecessary complexity.

Alternatively, we could just leave it as it is, or justify to not mark the function as unsafe because this a private function will only be used internally in the querying, and it is very unlikely to accidentally be used in an unsafe way. What do you think?

Reply:
> There are some possible workarounds to get rid of the unsafe functions in `ecs::filter`, for example: ...
> But I find workaround to be clunky and it is adding unnecessary complexity.

I see what you mean, though I do think it would make it clearer to readers what the intention behind the different implementations of the trait are doing. I could very easily see that `Read` implements the function `unsafe_fetch_parameter` while `And` implements `fetch_parameter` -- which tells me that one makes use of `unsafe` and the other does not. 

> Alternatively, we could just leave it as it is, or justify to not mark the function as unsafe because this a private function will only be used internally in the querying, and it is very unlikely to accidentally be used in an unsafe way.

It's true that it's currently only being used privately, but it _is_ an entirely public function in a public interface:
https://github.com/martinjonsson01/recs/blob/b189ebc8dc2b2174b6449a75d4d86e2e8abefe1c/crates/ecs/src/lib.rs#L361
Which can't be made private since it's being implemented in other modules (i.e. `ecs::filter`). 

But if it were to be private then I would still say that it's preferable to keep the (sometimes redundant) `unsafe` like you initially did. I'd rather have it unnecessarily `unsafe` than not marked `unsafe` at all when it sometimes is. And it's true, like you say, it's not as much of a problem since it's only used internally. How about we make it `pub(crate)`, and then leave it as-is?